cpp/src/arrow/builder.cc

```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+      break;
+    case 4:
+      *out = std::make_shared<Int32Array>(
+          int32(), length_, data_, null_bitmap_, null_count_);
+      break;
+    case 8:
+      *out = std::make_shared<Int64Array>(
+          int64(), length_, data_, null_bitmap_, null_count_);
+      break;
+    default:
+      DCHECK(false);
+      return Status::NotImplemented("Only ints of size 1,2,4,8 are supported");
+  }
+
+  data_ = null_bitmap_ = nullptr;
+  capacity_ = length_ = null_count_ = 0;
+  return Status::OK();
+}
+
+// Check if we would need to expand the underlying storage type
+uint8_t expanded_int_size(int64_t val, uint8_t current_int_size) {
+  if (current_int_size == 8 ||
+      (current_int_size < 8 &&
+          (val > static_cast<int64_t>(std::numeric_limits<int32_t>::max()) ||
+              val < static_cast<int64_t>(std::numeric_limits<int32_t>::min())))) {
+    return 8;
+  } else if (current_int_size == 4 ||
+             (current_int_size < 4 &&
+                 (val > static_cast<int64_t>(std::numeric_limits<int16_t>::max()) ||
+                     val < static_cast<int64_t>(std::numeric_limits<int16_t>::min())))) {
+    return 4;
+  } else if (current_int_size == 2 ||
+             (current_int_size == 1 &&
+                 (val > static_cast<int64_t>(std::numeric_limits<int8_t>::max()) ||
+                     val < static_cast<int64_t>(std::numeric_limits<int8_t>::min())))) {
+    return 2;
+  } else {
+    return 1;
+  }
+}
+
+Status AdaptiveIntBuilder::Append(
+    const int64_t* values, int64_t length, const uint8_t* valid_bytes) {
+  RETURN_NOT_OK(Reserve(length));
+
+  if (length > 0) {
+    if (int_size_ < 8) {
+      uint8_t new_int_size = int_size_;
+      for (int64_t i = 0; i < length; i++) {
+        if (valid_bytes == nullptr || valid_bytes[i]) {
+          new_int_size = expanded_int_size(values[i], new_int_size);
+        }
+      }
+      if (new_int_size != int_size_) { RETURN_NOT_OK(ExpandIntSize(new_int_size)); }
+    }
+  }
+
+  if (int_size_ == 8) {
+    std::memcpy(reinterpret_cast<int64_t*>(raw_data_) + length_, values,
+        sizeof(int64_t) * length);
+  } else {
+    // int_size_ may have changed, so we need to recheck
+    switch (int_size_) {
+      case 1: {
+        int8_t* data_ptr = reinterpret_cast<int8_t*>(raw_data_) + length_;
+        std::transform(values, values + length, data_ptr,
+            [](int64_t x) { return static_cast<int8_t>(x); });
+      } break;
+      case 2: {
+        int16_t* data_ptr = reinterpret_cast<int16_t*>(raw_data_) + length_;
+        std::transform(values, values + length, data_ptr,
+            [](int64_t x) { return static_cast<int16_t>(x); });
+      } break;
+      case 4: {
+        int32_t* data_ptr = reinterpret_cast<int32_t*>(raw_data_) + length_;
+        std::transform(values, values + length, data_ptr,
+            [](int64_t x) { return static_cast<int32_t>(x); });
+      } break;
+      default:
+        DCHECK(false);
```

Comment:
This code path should never be reached. Without the `default:` section, we would have a compiler warning about unhandled values.